chia/data_layer/data_store.py

```diff
@@ -821,3 +879,223 @@ async def get_operations(
                 previous_root = current_root
 
         return result
+
+    async def download_data(self, subscription: Subscription, target_hash: bytes32, *, lock: bool = True) -> bool:
+        tree_id = subscription.tree_id
+        ip = subscription.ip
+        port = int(subscription.port)
+        async with self.db_wrapper.locked_transaction(lock=lock):
+            exists = await self.tree_id_exists(tree_id, lock=False)
+            if not exists:
+                await self.create_tree(tree_id=tree_id, status=Status.COMMITTED, lock=False)
+            URL = f"http://{ip}:{port}/ws"
+            if subscription.mode is DownloadMode.LATEST:
+                async with aiohttp.ClientSession() as session:
+                    async with session.ws_connect(URL) as ws:
+                        request = {
+                            "type": "request_root",
+                            "tree_id": tree_id.hex(),
+                            "node_hash": target_hash.hex(),
+                        }
+                        await ws.send_str(json.dumps(request))
+                        msg = await ws.receive()
+                        root_json = json.loads(msg.data)
+                        node = root_json["node_hash"]
+                        internal_nodes = 0
+                        terminal_nodes = 0
+                        stack: List[str] = []
+                        add_to_db_cache: Dict[str, Any] = {}
+
+                        # TODO: Add back pagination. This needs historical ancestors.
+                        request = {
+                            "type": "request_nodes",
+                            "tree_id": tree_id.hex(),
+                            "node_hash": node,
+                        }
+                        await ws.send_str(json.dumps(request))
+                        msg = await ws.receive()
+                        msg_json = json.loads(msg.data)
+                        answer = msg_json["answer"]
+                        for row in answer:
+                            if row["is_terminal"]:
+                                key = row["key"]
+                                value = row["value"]
+                                hash = Program.to((hexstr_to_bytes(key), hexstr_to_bytes(value))).get_tree_hash()
+                                if hash.hex() == node:
+                                    await self._insert_node(
```

Comment:
switch to use  _insert_terminal_node and _insert_terminal_node and _insert_internal_node instead of _insert_node

Reply:
the root is inserted last - so worst case we'll have some orphan nodes, as no node will link any of them.

what concerns me more is calling the internal functions, instead of public ones. so I'm thinking to refactor it to a `insert_generation` and a cache, in this way at least we call them only in 1 place instead of all over the place.

Reply:
this code was intended to include pagination - i.e. for a big tree not fitting in memory, split it into multiple batches. then we won't be able to fit all of it into a cache. but I think we can worry about this later, for now I changed it to send the whole tree anyway.

Reply:
you will have orphan nodes which you might try to download again and then im not sure what will happen, maybe not in this pr but i dont think we should commit unvalidated nodes to the main tables

Reply:
looking into it we will skip values that are already in the node table and just do nothing but this should be the case in #10064 ancestor table as well